src/time.c

```diff
@@ -46,6 +46,43 @@ rcutils_time_point_value_as_nanoseconds_string(
   return RCUTILS_RET_OK;
 }
 
+rcutils_ret_t
+rcutils_time_point_value_as_date_string(
+  const rcutils_time_point_value_t * time_point,
+  char * str,
+  size_t str_size)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(time_point, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(str, RCUTILS_RET_INVALID_ARGUMENT);
+  if (0 == str_size) {
+    return RCUTILS_RET_OK;
+  }
+  // best to abs it to avoid issues with negative values in C89, see:
+  //   https://stackoverflow.com/a/3604984/671658
+  uint64_t abs_time_point = (uint64_t)llabs(*time_point);
+  // break into two parts to avoid floating point error
+  uint64_t seconds = abs_time_point / (1000u * 1000u * 1000u);
+  uint64_t nanoseconds = abs_time_point % (1000u * 1000u * 1000u);
+  char str2[21];   // Make sure the buffer is large enough to hold the largest possible uint64_t
+
+  if (rcutils_snprintf(str2, sizeof(str2), "%" PRIu64, nanoseconds)) {
+    RCUTILS_SET_ERROR_MSG("failed to format time point into string as iso8601_date");
+    return RCUTILS_RET_ERROR;
+  }
+
+  time_t now_t = (time_t)(seconds);
+  struct tm * ptm = localtime(&now_t);
+
+  strftime(str, 32, "%Y-%m-%d %H:%M:%S", ptm);
+  int position = 19;
+  if (rcutils_snprintf(&str[position], str_size - position, ".%.3s", str2)) {
```

Comment:
This doesn't seem to correspond to ISO 8601, now that I look at it: https://en.wikipedia.org/wiki/ISO_8601

That is, there is no provision in ISO 8601 to add in the nanoseconds to the date representation.  That's why there is no representation in `strftime` for it.

There are a couple of paths that we could go down here:
1.  Leave the representation with the nanoseconds applied to the end.  If we choose this, then I think we should rename the token to something like "date_time_with_ns".
2.  Change this to be iso8601 compliant.  If we do this, then I think the token should be named something like "iso8601_date_time".

(there is also the further problem that we don't know that the incoming `str` has enough room for the string, but let's deal with the output format first)

Reply:
I prefer: `date_time_with_ms` since I need the ms for my application